shotover-proxy/src/frame/cassandra.rs

```diff
@@ -241,52 +255,64 @@ impl CassandraFrame {
         })
     }
 
+    /// returns the query type for the current statement.
     pub fn get_query_type(&self) -> QueryType {
+        /*
+            Read,
+        Write,
+        ReadWrite,
+        SchemaChange,
+        PubSubMessage,
+             */
         match &self.operation {
-            CassandraOperation::Query {
-                query: CQL::Parsed(query),
-                ..
-            } => match query.as_ref() {
-                Statement::Query(_) => QueryType::Read,
-                Statement::Insert { .. } => QueryType::Write,
-                Statement::Update { .. } => QueryType::Write,
-                Statement::Delete { .. } => QueryType::Write,
-                // TODO: handle prepared, execute and schema change query types
-                _ => QueryType::Read,
-            },
+            CassandraOperation::Query { query: cql, .. } => {
+                // set to lowest type
+                let mut result = QueryType::SchemaChange;
+                for cql_statement in &cql.statements {
+                    result = match (cql_statement.get_query_type(), &result) {
+                        (QueryType::ReadWrite, _) => QueryType::ReadWrite,
+                        (QueryType::Write, QueryType::ReadWrite | QueryType::Write) => result,
+                        (QueryType::Write, QueryType::Read) => QueryType::ReadWrite,
+                        (QueryType::Write, QueryType::SchemaChange | PubSubMessage) => {
+                            QueryType::Write
+                        }
+                        (QueryType::Read, _) => {
+                            if result == QueryType::SchemaChange {
+                                QueryType::Read
+                            } else {
+                                result
+                            }
+                        }
+                        (QueryType::SchemaChange | PubSubMessage, _) => result,
+                    }
+                }
+                result
+            }
             _ => QueryType::Read,
         }
     }
 
-    pub fn namespace(&self) -> Vec<String> {
+    /// returns a list of table names from the CassandraOperation
+    pub fn get_table_names(&self) -> Vec<&FQName> {
```

Comment:
This should be implemented with self.queries as that will handle iterating over both QUERY and BATCH

Reply:
I was able to get the code compiling with:
```rust
    pub fn get_table_names_(&mut self) -> Vec<&FQName> {
        self.operation
            .queries()
            .into_iter()
            .filter_map(|stmt| CQLStatement::get_table_name(stmt))
            .collect()
    }
```

The compiler error reported that queries() was creating a temporary value that didnt live long enough, so I used into_iter so that the Vec returned by queries was consumed and no longer considered a temporary value.